src/serde/de_tree.rs

```diff
@@ -0,0 +1,327 @@
+use std::convert::TryInto;
+use std::io::{Error, Read, Result, Write};
+
+use sha2::Digest;
+
+use crate::sha2::Sha256;
+
+use super::parse_atom::decode_size_with_offset;
+use super::utils::{copy_exactly, skip_bytes};
+
+const MAX_SINGLE_BYTE: u8 = 0x7f;
+const CONS_BOX_MARKER: u8 = 0xff;
+
+struct ShaWrapper(Sha256);
+
+impl Write for ShaWrapper {
+    fn write(&mut self, blob: &[u8]) -> std::result::Result<usize, Error> {
+        self.0.update(blob);
+        Ok(blob.len())
+    }
+    fn flush(&mut self) -> std::result::Result<(), Error> {
+        Ok(())
+    }
+}
+
+/// This data structure is used with `parse_triples`, which returns a triple of
+/// integer values for each clvm object in a tree.
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ParsedTriple {
+    Atom {
+        start: u64,
+        end: u64,
+        atom_offset: u32,
+    },
+    Pair {
+        start: u64,
+        end: u64,
+        right_index: u32,
+    },
+}
+
+enum ParseOpRef {
+    ParseObj,
+    SaveCursor(usize),
+    SaveIndex(usize),
+}
+
+fn sha_blobs(blobs: &[&[u8]]) -> [u8; 32] {
+    let mut h = Sha256::new();
+    for blob in blobs {
+        h.update(blob);
+    }
+    h.finalize()
+        .as_slice()
+        .try_into()
+        .expect("wrong slice length")
+}
+
+fn tree_hash_for_byte(b: u8, calculate_tree_hashes: bool) -> Option<[u8; 32]> {
+    if calculate_tree_hashes {
+        Some(sha_blobs(&[&[1, b]]))
+    } else {
+        None
+    }
+}
+
+fn skip_or_sha_bytes<R: Read>(
+    f: &mut R,
+    skip_size: u64,
+    calculate_tree_hashes: bool,
+) -> Result<Option<[u8; 32]>> {
+    if calculate_tree_hashes {
+        let mut h = Sha256::new();
+        h.update([1]);
+        let mut w = ShaWrapper(h);
+        copy_exactly(f, &mut w, skip_size)?;
+        let r: [u8; 32] =
+            w.0.finalize()
+                .as_slice()
+                .try_into()
+                .expect("wrong slice length");
+        Ok(Some(r))
+    } else {
+        skip_bytes(f, skip_size)?;
+        Ok(None)
+    }
+}
+
+/// parse a serialized clvm object tree to an array of `ParsedTriple` objects
+
+/// This alternative mechanism of deserialization generates an array of
+/// references to each clvm object. A reference contains three values:
+/// a start offset within the blob, an end offset, and a third value that
+/// is either: an atom offset (relative to the start offset) where the atom
+/// data starts (and continues to the end offset); or an index in the array
+/// corresponding to the "right" element of the pair (in which case, the
+/// "left" element corresponds to the current index + 1).
+///
+/// Since these values are offsets into the original buffer, that buffer needs
+/// to be kept around to get the original atoms.
+
+type ParsedTriplesOutput = (Vec<ParsedTriple>, Option<Vec<[u8; 32]>>);
+
+pub fn parse_triples<R: Read>(
```

Comment:
The idea is to also support parsing from a stream without going through the stream twice (once to "pretend to parse" to turn it into a blob, then again parsing as a blob). Using an arbitrary `Read` object means you have to wrap it in another read object that saves the bytes into a `Vec<u8>` as they're requested. I believe by going through the stream just once, saving the read bytes along the way, you may end up being faster. The question is, how much does it negatively impact `Cursor` parses because you can't use `seek`.

Since these functions are parameterized by `R: Read`, my hope is that passing in a `Cursor` would cause of a lot of overhead to be optimized away. It's not clear to me how much of that is *actually* done, but I think it's possible in theory. Can the compiler recognize that copying a `Cursor` to a [`sink`](https://doc.rust-lang.org/std/io/struct.Sink.html) can just do a `seek`-type op? Maybe it's a stretch.

Reply:
> The idea is to also support parsing from a stream without going through the stream twice

Do you have a use case for that? (if not, I'd say: ["you ain't gonna need it"](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it))

Reply:
My past experiences suggest that not being able to deserialize from a stream is an oversight.

Reply:
I don't think we've come across a use for parsing clvm from a stream with our existing functions. partly because it's unfeasible to bind a python stream to be used as a rust stream. but even on the python side we always use buffers.

Reply:
> I'm a bit more unclear on what the use case for computing all the tree hashes is. We typically only want tree hashes at the curry/uncurry boundaries, I think.

How do we know where those boundaries are at parse time?